domain/permission/state/shared.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	corepermission "github.com/juju/juju/core/permission"
+	permissionerrors "github.com/juju/juju/domain/permission/errors"
+	internaldatabase "github.com/juju/juju/internal/database"
+)
+
+// SharedState holds the shared state for the permission domain. Methods
+// which are shared between multiple domains.
+type SharedState struct {
+	statementBase internaldatabase.StatementBase
+}
+
+// NewSharedState creates a new SharedState.
+func NewSharedState(base internaldatabase.StatementBase) *SharedState {
+	return &SharedState{statementBase: base}
+}
+
+// AddUserPermissionArgs is a specification for adding a user permission.
+type AddUserPermissionArgs struct {
+	PermissionUUID string
+	UserUUID       string
+	Access         corepermission.Access
+	Target         corepermission.ID
+}
+
+// AddUserPermission adds a permission for the given user on the given target.
+func (s SharedState) AddUserPermission(ctx context.Context, tx *sqlair.TX, spec AddUserPermissionArgs) error {
```

Comment:
I'm rethinking the contents of this method. It assumes that the caller has made all of the necessary checks before doing the insert. The uniqueness of the permission is guaranteed by the sql defining the permission table. But the user nor the target are guaranteed. 

In the one case this method is called outside of the permission domain, the user is being created. I see errors cropping up with other users.

Reply:
I'm just moving code here. We can address the code in these methods in a follow PR.